lib/column/variant.go

```diff
@@ -0,0 +1,385 @@
+}
+
+func (c *Variant) Name() string {
+	return c.name
+}
+
+func (c *Variant) Type() Type {
+	return c.chType
+}
+
+func (c *Variant) Rows() int {
+	return len(c.discriminators)
+}
+
+func (c *Variant) Row(i int, ptr bool) any {
+	typeIndex := c.discriminators[i]
+	offsetIndex := c.offsets[i]
+	var value any
+	var chType string
+	if typeIndex != NullVariantDiscriminator {
+		value = c.columns[typeIndex].Row(offsetIndex, ptr)
+		chType = string(c.columns[typeIndex].Type())
+	}
+
+	vt := chcol.NewVariantWithType(value, chType)
+	if ptr {
+		return &vt
+	}
+
+	return vt
+}
+
+func (c *Variant) ScanRow(dest any, row int) error {
+	typeIndex := c.discriminators[row]
+	offsetIndex := c.offsets[row]
+	var value any
+	var chType string
+	if typeIndex != NullVariantDiscriminator {
+		value = c.columns[typeIndex].Row(offsetIndex, false)
+		chType = string(c.columns[typeIndex].Type())
+	}
+
+	switch v := dest.(type) {
+	case *chcol.Variant:
+		vt := chcol.NewVariant(value)
+		*v = vt
+	case **chcol.Variant:
+		vt := chcol.NewVariant(value)
+		**v = vt
+	case *chcol.VariantWithType:
+		vt := chcol.NewVariantWithType(value, chType)
+		*v = vt
+	case **chcol.VariantWithType:
+		vt := chcol.NewVariantWithType(value, chType)
+		**v = vt
+	default:
+		if typeIndex == NullVariantDiscriminator {
+			return nil
+		}
+
+		if err := c.columns[typeIndex].ScanRow(dest, offsetIndex); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (c *Variant) Append(v any) (nulls []uint8, err error) {
+	switch v.(type) {
+	case []chcol.Variant:
+		for i, vt := range v.([]chcol.Variant) {
+			err := c.AppendRow(vt)
+			if err != nil {
+				return nil, fmt.Errorf("failed to AppendRow at index %d: %w", i, err)
+			}
+		}
+
+		return nil, nil
+	case []*chcol.Variant:
+		for i, vt := range v.([]*chcol.Variant) {
+			err := c.AppendRow(vt)
+			if err != nil {
+				return nil, fmt.Errorf("failed to AppendRow at index %d: %w", i, err)
+			}
+		}
+
+		return nil, nil
+	case []chcol.VariantWithType:
+		for i, vt := range v.([]chcol.VariantWithType) {
+			err := c.AppendRow(vt)
+			if err != nil {
+				return nil, fmt.Errorf("failed to AppendRow at index %d: %w", i, err)
+			}
+		}
+
+		return nil, nil
+	case []*chcol.VariantWithType:
+		for i, vt := range v.([]*chcol.VariantWithType) {
+			err := c.AppendRow(vt)
+			if err != nil {
+				return nil, fmt.Errorf("failed to AppendRow at index %d: %w", i, err)
+			}
+		}
+
+		return nil, nil
+	default:
+		if valuer, ok := v.(driver.Valuer); ok {
+			val, err := valuer.Value()
+			if err != nil {
+				return nil, &ColumnConverterError{
+					Op:   "Append",
+					To:   string(c.chType),
+					From: fmt.Sprintf("%T", v),
+					Hint: "could not get driver.Valuer value",
+				}
+			}
+
+			return c.Append(val)
+		}
+
+		return nil, &ColumnConverterError{
+			Op:   "Append",
+			To:   string(c.chType),
+			From: fmt.Sprintf("%T", v),
+		}
+	}
+}
+
+func (c *Variant) AppendRow(v any) error {
+	var requestedType string
+	switch v.(type) {
```

Comment:
```suggestion
	switch vv := v.(type) {
```

This will give you a type asserted value ready to use inside case blocks without the need for additional type assertions.

Reply:
Ah I've seen this elsewhere in the code, good point. I can simplify this switch